### PR TITLE
fix self test

### PR DESCRIPTION
**Pull request by InternetofAwesome**
ST's docs are pretty bad on the self test features. 
Basically, you need to do the difference between the two self test modes.

This should be right. Will double check with ST, but as I recall from the work I did a few years ago on a similar part, this is right. Also, now it passes for a part that we have observed to be good.

### CODE REVIEW

**Comment by InternetofAwesome** [comment]
One more modification, from referencing ST's drivers:
https://github.com/STMicroelectronics/STMems_Linux_IIO_drivers/blob/linux-4.4.y-gh/drivers/iio/accel/st_lis2dw12_core.c#L810

**Comment by InternetofAwesome** [comment]
Ok. Ready for review/merge

File: hw/drivers/sensors/lis2dw12/src/lis2dw12.c [hunk 1]
```diff
@@ -1909,14 +1909,11 @@ int lis2dw12_run_self_test(struct sensor_itf *itf, int *result)
...
     for(i = 0; i < 3; i++) {
-        change += pos[i] - base[i];
-    }
-
-    if ((change > 70) && (change < 1500)) {
-        *result = 0;
-    } else {
-        *result = -1;
+    	change = pos[i] - base[i];
+    	if(!(change > 70) && (change < 1500))
```
Comment (mkiiskila) [inline hw/drivers/sensors/lis2dw12/src/lis2dw12.c]: Coding style: need whitespace after 'if'. More importantly: even single line statements after if-conditions must be wrapped with { }.

**Comment by vrahane** [comment]
Spoke to @InternetofAwesome. This needs some more work.

**Comment by InternetofAwesome** [comment]
Allright, ready for review

File: hw/drivers/sensors/lis2dw12/src/lis2dw12.c [hunk 2]
```diff
@@ -1863,60 +1915,94 @@ int lis2dw12_get_int1_on_int2_map(struct sensor_itf *itf, uint8_t *val)
...
     }
 
-    /* calculate accel data difference */
-    change = 0;
-    for(i = 0; i < 3; i++) {
-        change += pos[i] - base[i];
-    }
+    /* take negative offset reading */
+    for(i=0; i<5; i++)
+        {
```
Comment (vrahane) [inline hw/drivers/sensors/lis2dw12/src/lis2dw12.c]: coding_style: Opening brace should be on the same line as the `for`.

File: hw/drivers/sensors/lis2dw12/src/lis2dw12.c [hunk 3]
```diff
@@ -1863,60 +1915,94 @@ int lis2dw12_get_int1_on_int2_map(struct sensor_itf *itf, uint8_t *val)
...
+
+    /* disable self test mod */
+    rc = lis2dw12_writelen(itf, LIS2DW12_REG_CTRL_REG1, prev_config, 6);
+        if (rc) {
+            return rc;
+        }
+
+    /* compare values to thresholds */
+    *result = 0;
+    for(i = 0; i < 3; i++) {
```
Comment (vrahane) [inline hw/drivers/sensors/lis2dw12/src/lis2dw12.c]: coding_style: A space is needed after `for`

File: hw/drivers/sensors/lis2dw12/src/lis2dw12.c [hunk 4]
```diff
@@ -1863,60 +1915,94 @@ int lis2dw12_get_int1_on_int2_map(struct sensor_itf *itf, uint8_t *val)
...
+    	if((diff[i] < min) || (diff[i] > max)) {
```
Comment (vrahane) [inline hw/drivers/sensors/lis2dw12/src/lis2dw12.c]: coding_style: A space is needed after `if`

File: hw/drivers/sensors/lis2dw12/src/lis2dw12.c [hunk 5]
```diff
@@ -1863,60 +1915,94 @@ int lis2dw12_get_int1_on_int2_map(struct sensor_itf *itf, uint8_t *val)
...
+    		*result -= 1;
```
Comment (vrahane) [inline hw/drivers/sensors/lis2dw12/src/lis2dw12.c]: coding_style: There is a tab here probably, what you need here is 4 spaces.

File: hw/drivers/sensors/lis2dw12/src/lis2dw12.c [hunk 6]
```diff
@@ -1863,60 +1915,94 @@ int lis2dw12_get_int1_on_int2_map(struct sensor_itf *itf, uint8_t *val)
...
+    {
+
+        rc = lis2dw12_get_data(itf, 2, &(data[0]), &(data[1]), &(data[2]));
+        if (rc) {
+            return rc;
+        }
+        diff[0] += data[0];
+        diff[1] += data[1];
+        diff[2] += data[2];
+        //wait at least 20 ms
```
Comment (vrahane) [inline hw/drivers/sensors/lis2dw12/src/lis2dw12.c]: coding_style: C++ comments are not allowed.

File: hw/drivers/sensors/lis2dw12/src/lis2dw12.c [hunk 7]
```diff
@@ -1863,60 +1915,94 @@ int lis2dw12_get_int1_on_int2_map(struct sensor_itf *itf, uint8_t *val)
...
+     * both negative and positive offset */
+    int min = LIS2DW12_ST_MIN*5*2;
+    int max = LIS2DW12_ST_MAX*5*2;
+
+    int16_t data[3], diff[3] = {0,0,0};
     int i;
-    int16_t change;
-    uint8_t fs;
+    uint8_t prev_config[6];
+    //set config per datasheet, with positive self test mode enabled.
```
Comment (vrahane) [inline hw/drivers/sensors/lis2dw12/src/lis2dw12.c]: coding_style: C++ comments are not allowed.

**Comment by vrahane** [comment]
@InternetofAwesome is this ready to review ?

**Comment by InternetofAwesome** [comment]
Ok. Ready for review.

File: hw/drivers/sensors/lis2dw12/src/lis2dw12.c [hunk 8]
```diff
@@ -1863,60 +1915,92 @@ int lis2dw12_get_int1_on_int2_map(struct sensor_itf *itf, uint8_t *val)
...
+    int min = LIS2DW12_ST_MIN*5*2;
+    int max = LIS2DW12_ST_MAX*5*2;
+
+    int16_t data[3], diff[3] = {0,0,0};
     int i;
-    int16_t change;
-    uint8_t fs;
+    uint8_t prev_config[6];
+    /* set config per datasheet, with positive self test mode enabled. */
+    uint8_t st_config[] = {0x44, 0x08, 0x40, 0x00, 0x00, 0x10};
```
Comment (jacobrosenthal) [inline hw/drivers/sensors/lis2dw12/src/lis2dw12.c]: Would be nice if this was explicit or at least documented in comments what it was trying to do. For instance youre enabling self test here, and then again later...Is that purposeful? Also you seem to set FS_4G here. Any particular reasoning? In many other implementations including the one you linked I find 2g

File: hw/drivers/sensors/lis2dw12/src/lis2dw12.c [hunk 9]
```diff
@@ -1863,60 +1915,92 @@ int lis2dw12_get_int1_on_int2_map(struct sensor_itf *itf, uint8_t *val)
...
+    rc = lis2dw12_set_self_test(itf, LIS2DW12_ST_MODE_MODE1);
     if (rc) {
         return rc;
     }
 
-    /* set self test mode to positive self test */
-    rc = lis2dw12_set_self_test(itf, LIS2DW12_ST_MODE_MODE1);
+    /* wait 100ms */
+    os_time_delay(OS_TICKS_PER_SEC / 100);
+    rc = lis2dw12_get_data(itf, 2, &(data[0]), &(data[1]), &(data[2]));
```
Comment (jacobrosenthal) [inline hw/drivers/sensors/lis2dw12/src/lis2dw12.c]: Should not be using hardcoded FS value, but rather the FS_4G define, at all the other lis2dw12_get_data calls too

File: hw/drivers/sensors/lis2dw12/src/lis2dw12.c [hunk 10]
```diff
@@ -1863,60 +1915,92 @@ int lis2dw12_get_int1_on_int2_map(struct sensor_itf *itf, uint8_t *val)
...
+    /* disable self test mod */
+    rc = lis2dw12_writelen(itf, LIS2DW12_REG_CTRL_REG1, prev_config, 6);
+        if (rc) {
+            return rc;
+        }
+
+    /* compare values to thresholds */
+    *result = 0;
+    for (i = 0; i < 3; i++) {
+        if ((diff[i] < min) || (diff[i] > max)) {
```
Comment (jacobrosenthal) [inline hw/drivers/sensors/lis2dw12/src/lis2dw12.c]: My device fails here on negative values, and the st driver you linked does do an abs

Comment (jacobrosenthal) [reply hw/drivers/sensors/lis2dw12/src/lis2dw12.c]: (gdb) p diff   
$5 = {-3419, -3419, -3419}

(gdb) p diff   
$10 = {-1191, -1878, -1882}

And with an abs Im still failing low occasionally, so Im not sure on the methadology.

(gdb) p diff
$10 = {63, 63, 63}

**Pull request merged**